### ISSUE TITLE
Pi05 output mismatch between PyTorch and Jax weights

### ISSUE THREAD

**Issue by littlespray** [Physical-Intelligence/openpi#810]
Hi all,

When testing with the following script, I found that the output of pi05’s weight before and after conversion is not the same given identical input. May I please ask what could be the possible reason?

I used the conversion command provided and after conversion, I copied the ``assets`` folder of the JAX weight to ``pi05_libero_pytorch/assets``; otherwise, it would report a ``Norm stat not found`` error.

```bash
uv run examples/convert_jax_model_to_pytorch.py \
    --checkpoint_dir ~/.cache/openpi/openpi-assets/checkpoints/pi05_libero \
    --config_name pi05_libero \
    --output_path ~/.cache/openpi/openpi-assets/checkpoints/pi05_libero_pytorch \
```

The script is:

```python
import numpy as np
from openpi.policies import libero_policy
from openpi.policies import policy_config as _policy_config
from openpi.shared import download
from openpi.training import config as _config

config = _config.get_config("pi05_libero")
checkpoint_dir = download.maybe_download("gs://openpi-assets/checkpoints/pi05_libero")

policy = _policy_config.create_trained_policy(config, checkpoint_dir)
example = libero_policy.make_libero_example()
result = policy.infer(example)
jax_actions = np.array(result["actions"])
del policy

pytorch_policy = _policy_config.create_trained_policy(config, "~/.cache/openpi/openpi-assets/checkpoints/pi05_libero_pytorch")
pytorch_result = pytorch_policy.infer(example)
pytorch_actions = np.array(pytorch_result["actions"])
del pytorch_policy

diff = jax_actions - pytorch_actions
max_diff = np.max(np.abs(diff))
mean_diff = np.mean(np.abs(diff))

print(f"Max Absolute Difference: {max_diff}")
print(f"Mean Absolute Difference: {mean_diff}")

if max_diff < 1e-5:
    print("The results are effectively identical.")
else:
    print("There are significant differences between the results.")
```

**Comment by flowmatching** [comment]
@littlespray I noticed that you closed this issue. Was the problem resolved? Are the outputs of the JAX and PyTorch versions still inconsistent for the same input? If so, did you find a solution?

**Comment by zttiannnn** [comment]
I met the same problem. Could you share how you resolved it? Really appreciate your help!

**Comment by ybwowen** [comment]
```
Max Absolute Difference: 2.0249355556025117
Mean Absolute Difference: 0.0860363834142548
There are significant differences between the results.
```
Here is my result. Is it normal that there is a significant numerical difference?

**Comment by littlespray** [comment]
Hi, 

The weight of the transformation is fine, but noise and random seed (Pytorch & **JAX**) need to be fixed when testing.

**Comment by ybwowen** [comment]
@littlespray Thanks for the hint! Could you please share a snippet of how you fixed the seed and noise for both JAX and PyTorch in the context of openpi?

**Comment by Agenda8** [comment]
@ybwowen  Hi, I'm using the JAX version and find the evaluation results unstable. Have you solved this? How should I fix the seed and noise for JAX? Thanks.

**Comment by ybwowen** [comment]
> [@ybwowen](https://github.com/ybwowen) Hi, I'm using the JAX version and find the evaluation results unstable. Have you solved this? How should I fix the seed and noise for JAX? Thanks.

Hi! It turns out that I was using wrong precision of the pytorch pretrained weight. Set `pytorch_training_weight="float32"` in the config resolves the problem.

**Comment by yangfansun** [comment]
> > [@ybwowen](https://github.com/ybwowen) Hi, I'm using the JAX version and find the evaluation results unstable. Have you solved this? How should I fix the seed and noise for JAX? Thanks.
> 
> Hi! It turns out that I was using wrong precision of the pytorch pretrained weight. Set `pytorch_training_weight="float32"` in the config resolves the problem.

@ybwowen 
Hello, I didn't find any parameter named pytorch_training_weight in the project. Could you tell me where this was configured? Also, when you say "solving the problem," do you mean that the model inference results match exactly, or that the MAE metric drops to a very low level, such as 1e-4 or 1e-5? Thank you.

**Comment by ybwowen** [comment]
> > > [@ybwowen](https://github.com/ybwowen) Hi, I'm using the JAX version and find the evaluation results unstable. Have you solved this? How should I fix the seed and noise for JAX? Thanks.
> > 
> > 
> > Hi! It turns out that I was using wrong precision of the pytorch pretrained weight. Set `pytorch_training_weight="float32"` in the config resolves the problem.
> 
> [@ybwowen](https://github.com/ybwowen) Hello, I didn't find any parameter named pytorch_training_weight in the project. Could you tell me where this was configured? Also, when you say "solving the problem," do you mean that the model inference results match exactly, or that the MAE metric drops to a very low level, such as 1e-4 or 1e-5? Thank you.

Hi! Sorry for the mistake, the correct parameter name is `pytorch_training_precision`. I did not compare the model inference results later, but it seems that the performance is good, and I got the expected results on benchmarks such as libero.
